src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.AvoidUnusedMembers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class AvoidUnusedMembersDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        internal static string UnunsedMemberNameProperty = nameof(UnunsedMemberNameProperty);
+        internal static string UnunsedMemberKindProperty = nameof(UnunsedMemberKindProperty);
+
+        private readonly DiagnosticDescriptor _noReadsUnnecessaryWithSuggestionDescriptor;
+
+        public AvoidUnusedMembersDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.AvoidUnusedMembersDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _noReadsUnnecessaryWithSuggestionDescriptor = new DiagnosticDescriptor(
+                UnnecessaryWithSuggestionDescriptor.Id,
+                UnnecessaryWithSuggestionDescriptor.Title,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Members_Without_Reads_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                UnnecessaryWithSuggestionDescriptor.Category,
+                UnnecessaryWithSuggestionDescriptor.DefaultSeverity,
+                UnnecessaryWithSuggestionDescriptor.IsEnabledByDefault,
+                UnnecessaryWithSuggestionDescriptor.Description,
+                UnnecessaryWithSuggestionDescriptor.HelpLinkUri,
+                UnnecessaryWithSuggestionDescriptor.CustomTags.ToArray());
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // We want to analyze references in generated code, but not report unused members in generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
```

Comment:
please follow IDE conventions.  We do not do the 'ultra nested function' approach to analyzers.  We sometimes hav ea small amount of code here (for example to get compilation information, or to create some shared state).  but then we call explicit functions that are passed this information, rather than having nested functions that capture.  it makes *data flow* explicit and helps with function readability.  I know you like this style, but it stands out highly against all the other features and I would prefer we not unnecesarily do that.